File: sources/src/main/csharp/Graph/BreadthFirstSearch.cs
namespace Ssw.Archive.Graph;

/// <summary>Traverses the source-reachable directed graph in breadth-first order.</summary>
/// <remarks>
/// Vertex IDs are ordered unique strings, source and every edge endpoint are declared, and outgoing neighbors follow
/// edge input order. A vertex is emitted once when first reached; unreachable vertices are omitted. Inputs are not
/// mutated.
/// source를 queue에 넣고 현재 정점의 미방문 이웃을 차례로 enqueue해 간선 수 기준의 층을 넓혀 가요.
/// 정점은 enqueue 순간 방문 표시되어 정확히 한 번만 결과와 queue에 들어가는 불변식을 유지해요.
/// 이웃은 간선 입력 순서를 따르므로 출력이 결정적이며 선언되지 않은 정점·중복 ID는 탐색 전에 거부해요.
/// 시간 복잡도는 O(V + E), 인접 목록과 방문·queue 공간은 O(V + E)예요.
/// </remarks>
public static class BreadthFirstSearch {
    /// <summary>Represents one directed edge.</summary>
    public sealed record Edge(string From, string To);

    /// <summary>Returns source-reachable vertices in deterministic breadth-first order.</summary>
    public static IReadOnlyList<string> Execute(
        IReadOnlyList<string> vertices,
        IReadOnlyList<Edge> edges,
        string source) {
        var adjacency = BuildAdjacency(vertices, edges);
        ArgumentNullException.ThrowIfNull(source);
        if (!adjacency.ContainsKey(source)) throw new ArgumentException("Source is not declared.", nameof(source));
        var visited = new HashSet<string>(StringComparer.Ordinal) { source };
        var queue = new List<string> { source };
        for (var cursor = 0; cursor < queue.Count; cursor++) {
            foreach (var next in adjacency[queue[cursor]]) if (visited.Add(next)) queue.Add(next);
        }
        return queue;
    }

    private static Dictionary<string, List<string>> BuildAdjacency(
        IReadOnlyList<string> vertices,
        IReadOnlyList<Edge> edges) {
        ArgumentNullException.ThrowIfNull(vertices);
        ArgumentNullException.ThrowIfNull(edges);
        var adjacency = new Dictionary<string, List<string>>(StringComparer.Ordinal);
        foreach (var vertex in vertices) {
            ArgumentNullException.ThrowIfNull(vertex);
            if (!adjacency.TryAdd(vertex, [])) throw new ArgumentException("Vertex IDs must be unique.", nameof(vertices));
        }
        foreach (var edge in edges) {
            ArgumentNullException.ThrowIfNull(edge);
            if (!adjacency.ContainsKey(edge.From) || !adjacency.ContainsKey(edge.To)) {
                throw new ArgumentException("Edge endpoint is not declared.", nameof(edges));
            }
            adjacency[edge.From].Add(edge.To);
        }
        return adjacency;
    }
}

File: sources/src/main/javascript/graph/breadth-first-search.js
"use strict";

/**
 * 시작 정점에서 큐로 거리를 한 층씩 넓혀 방향 그래프의 도달 정점을 너비 우선으로 방문해요.
 * 정점은 처음 발견될 때 한 번만 큐에 넣어 각 방문 기록이 최단 간선 수 층에 속하도록 유지해요.
 * 선언된 정점 순서로 인접 정점을 처리하며 중복 정점·잘못된 끝점·시작점 오류를 순회 전에 거부해요.
 * 정점 수 V와 간선 수 E에 대해 시간은 O(V + E), 추가 공간은 O(V)예요.
 */
/**
 * @param {string} code
 * @param {string} [message]
 * @returns {never}
 */
function fail(code, message = code) {
  const error = new Error(message);
  error.code = code;
  throw error;
}

/**
 * Breadth-first traversal order from a declared source on a directed graph.
 *
 * Domain public API for `breadth-first-search`. JSON contract adapters live in the runner.
 * Vertex IDs are ordered unique strings, source and every edge endpoint are declared, and
 * outgoing neighbors follow edge input order. A vertex is emitted once when first reached;
 * unreachable vertices are omitted. Inputs are not mutated.
 *
 * @param {readonly string[]} vertices
 * @param {readonly {from: string, to: string}[]} edges
 * @param {string} source
 * @returns {string[]}
 */
function execute(vertices, edges, source) {
  if (!Array.isArray(vertices) || vertices.some(vertex => typeof vertex !== "string")) {
    fail("INVALID_INPUT", "vertices must be an array of strings");
  }
  /** @type {Map<string, string[]>} */
  const adjacency = new Map();
  for (const vertex of vertices) {
    if (adjacency.has(vertex)) fail("INVALID_INPUT", "vertex IDs must be unique");
    adjacency.set(vertex, []);
  }
  if (typeof source !== "string" || !adjacency.has(source)) {
    fail("INVALID_SOURCE", "source must be one of vertices");
  }
  if (!Array.isArray(edges)) fail("INVALID_EDGE", "edges must be an array");
  for (const edge of edges) {
    if (
      !edge
      || typeof edge.from !== "string"
      || typeof edge.to !== "string"
      || !adjacency.has(edge.from)
      || !adjacency.has(edge.to)
    ) {
      fail("INVALID_EDGE", "edge endpoint must be in vertices");
    }
    adjacency.get(edge.from).push(edge.to);
  }

  const visited = new Set([source]);
  /** @type {string[]} */
  const queue = [source];
  /** @type {string[]} */
  const order = [];
  for (let head = 0; head < queue.length; head++) {
    const current = queue[head];
    order.push(current);
    for (const next of adjacency.get(current)) {
      if (!visited.has(next)) {
        visited.add(next);
        queue.push(next);
      }
    }
  }
  return order;
}

module.exports = { execute };
